opm/simulators/aquifers/AquiferConstantFlux.hpp

```diff
@@ -0,0 +1,147 @@
+        : AquiferInterface<TypeTag>(aquifer.id, ebos_simulator)
+         , connections_(connections)
+         , aquifer_data_(aquifer)
+    {
+        // init_cumulative_flux is the flux volume from previoius running
+        this->initializeConnections();
+        connection_flux_.resize(this->connections_.size(), {0});
+    }
+
+    virtual ~AquiferConstantFlux() = default;
+
+    void updateAquifer(const SingleAquiferFlux& aquifer) {
+        aquifer_data_ = aquifer;
+    }
+
+    void initFromRestart(const data::Aquifers& /* aquiferSoln */) {
+    }
+
+    void initialSolutionApplied() {
+    }
+
+    void beginTimeStep() {
+    }
+
+    void endTimeStep() {
+        this->flux_rate_ = 0.;
+        for (const auto& q : this->connection_flux_) {
+            this->flux_rate_ += Opm::getValue(q);
+        }
+
+        this->cumulative_flux_ += this->flux_rate_ * this->ebos_simulator_.timeStepSize();
+    }
+
+    data::AquiferData aquiferData() const
+    {
+        data::AquiferData data;
+        data.aquiferID = this->aquifer_data_.id;
+        // pressure for constant flux aquifer is 0
+        data.pressure = 0.;
+        data.fluxRate = 0.;
+        for (const auto& q : this->connection_flux_) {
+            data.fluxRate += q.value();
+        }
+        data.volume = this->cumulative_flux_;
+        // not totally sure whether initPressure matters
+        data.initPressure = 0.;
+        return data;
+    }
+
+    void addToSource(RateVector& rates,
+                     const unsigned cellIdx,
+                     const unsigned timeIdx) {
+        const auto& model = this->ebos_simulator_.model();
+
+        const int idx = this->cellToConnectionIdx_[cellIdx];
+        if (idx < 0)
+            return;
+
+        const auto* intQuantsPtr = model.cachedIntensiveQuantities(cellIdx, timeIdx);
+        if (intQuantsPtr == nullptr) {
+            throw std::logic_error("Invalid intensive quantities cache detected in AquiferAnalytical::addToSource()");
+        }
+
+        const double fw = this->aquifer_data_.flux;
+        this->connection_flux_[idx] = fw * this->connections_[idx].effective_facearea;
+        rates[BlackoilIndices::conti0EqIdx + compIdx_()]
+                += this->connection_flux_[idx] / model.dofTotalVolume(cellIdx);
+    }
+
+private:
+    const std::vector<Aquancon::AquancCell>& connections_;
+    SingleAquiferFlux aquifer_data_;
+    std::vector<int> cellToConnectionIdx_;
+    std::vector<Eval> connection_flux_;
+    double flux_rate_ {};
+    double cumulative_flux_ = 0.;
+
+    void initializeConnections() {
+        this->cellToConnectionIdx_.resize(this->ebos_simulator_.gridView().size(/*codim=*/0), -1);
+        for (std::size_t idx = 0; idx < this->connections_.size(); ++idx) {
+            const auto global_index = this->connections_[idx].global_index;
+            const int cell_index = this->ebos_simulator_.vanguard().compressedIndexForInterior(global_index);
+
+            if (cell_index < 0) {
```

Comment:
this `if` statement might not be needed, since  a negative cell_index will serve its purpose to say that the cell is not connected to aquifer.

Reply:
> this `if` statement might not be needed, since a negative cell_index will serve its purpose to say that the cell is not connected to aquifer.

Sure, but without **some kind** of condition here you'll index out of bounds in `cellToConnectionIdx_`.  E.g., you'll generate the equivalent of
```C++
this->cellToConnectionIdx_[-1] = idx
```
and that's going to be a problem

Reply:
You are right, I misread it. Thanks.